resources/deb/postinst.erb

```diff
@@ -44,6 +44,27 @@ fi
   <%- end -%>
 <%- end -%>
 
+# adding support for triggers
+<%- install_triggers = get_interest_triggers("install").uniq
+  upgrade_triggers = get_interest_triggers("upgrade").uniq -%>
```

Comment:
Can the logic from lines 48 and 49 be moved out of the ERB and into methods in one of the underlying classes? Logic beyond "if/else" in an ERB is a code smell.

Reply:
Is it the `uniq` you want moved out? I think that could be moved into the `get_interest_triggers` method, but mostly this is just setting variables that we can use later on to save calling this method every time, which seems ok to me :woman_shrugging:

Reply:
No, I mean that the assignment should move out of the ERB. ERB files are templates, so they shouldn't have assignment in them -- ideally they're just mapping data from an instance of something to a human-readable format. In this case by using memoization, the assignment could be shoved out of the template and into the underlying class with (I believe) the same runtime cost (comments inline):

```ruby
class Vanagon
  class Project
    # Renamed to a bang (!) method, because it's always live and it never caches
    def get_interest_triggers!(pkg_state)
      interest_triggers = []
      check_pkg_state_string(pkg_state)
      components.flat_map(&:interest_triggers).compact.select { |s| s.pkg_state.include? pkg_state }
      interests.each do |interest|
        interest_triggers.push(interest)
      end
      # I am of the opinion that this many transformations chained together
      # (6 as of last count) is a code smell -- maybe flag this for future
      # refactoring in the comments?
      interest_triggers.flatten.uniq.compact
    end

    def get_interest_triggers(pkg_state)
      # Ensure that @interest_triggers is always a Hash
      @interest_triggers ||= {}
      # Now, check the hash to see if we've cached the computed triggers for
      # a given value of pkg_state (hashed for efficiency); if it has been,
      # then just return it because we don't need to keep building that Hash
      return @interest_triggers if @interest_triggers[pkg_state.hash]
      # Otherwise, assign the computed list of triggers with the hashed
      # value of pkg_state as a key; I believe this should return the
      # expected computed interest_triggers
      @interest_triggers[pkg_state.hash] = get_interest_triggers!(pkg_state)
    end

    def get_install_triggers
      get_interest_triggers("install")
    end

    def get_upgrade_triggers
      get_interest_triggers("upgrade")
    end
  end
end
```

Then `#get_install_triggers` and `#get_upgrade_triggers` can be called anywhere in the ERB template without concern for the cost of computing them or the need to worry about state or assignment in the ERB template itself.

Reply:
(`#get_install_triggers` and `#get_upgrade_triggers` could be renamed to just `#install_triggers`/`#upgrade_triggers`and that'd be a drop-in for the variables as named now, I think)